### PR TITLE
use `Object.defineProperty` in priority

### PR DESCRIPTION
**Pull request by dickeylth**
in case of `Uncaught TypeError: Cannot assign to read only property 'toString' of object '[object Object]'` under strict mode.

### CODE REVIEW

**Comment by lpinca** [comment]
We can't use `Object.defineProperty()` as the code should work also in very old browsers which only support ES3.

**Comment by dickeylth** [comment]
Let me explain the reason for the pr.
Our code runs in [weex](https://github.com/alibaba/weex), which is much similar with react-native, a framework for building mobile cross-platform UI, and the javascript codes run in JavaScriptCore/V8 runtime, in lately version in the js framework part the original object's prototype is freezed, so that overriding `toString` method could only be realized by `Object.defineProperty`.

File: index.js [hunk 1]
```diff
@@ -344,7 +344,12 @@ URL.prototype.toString = function toString(stringify) {
   if (url.hash) result += url.hash;
 
   return result;
-};
+}
+if (Object.defineProperty) {
+  Object.defineProperty(URL.prototype, 'toString', { value: toString });
```
Comment (lpinca) [inline index.js]: `writable`, `enumerable`, and `configurable` should be `true` in the descriptor.

File: index.js [hunk 2]
```diff
@@ -344,7 +344,17 @@ URL.prototype.toString = function toString(stringify) {
   if (url.hash) result += url.hash;
 
   return result;
-};
+}
+if (Object.defineProperty) {
```
Comment (lpinca) [inline index.js]: Can you please add a blank line before this one?

**Comment by lpinca** [comment]
Question is: isn't this a bug in weex? We are assigning to the URL prototype, not the Object prototype.

**Comment by dickeylth** [comment]
It seems that it has freeze `prototype`: https://github.com/alibaba/weex/pull/1529/files#diff-5f27557ff70d8602cacd1b1a065e63d7, and once the prototype is frozen, the instance can't override its properties.

**Comment by lpinca** [comment]
@dickeylth instead of using `Object.defineProperty()` how about this?

```js
function set(...) {...}
function toString() {...}

URL.prototype = { set: set, toString: toString };
```

it should work.

**Comment by dickeylth** [comment]
@lpinca You're right, it did work, I'll modify the commit.

**Comment by lpinca** [comment]
Still LGTM. I'll merge this later today or tomorrow if other members don't do it first.

**Comment by dickeylth** [comment]
@lpinca Thank you for your patience!

**Comment by lpinca** [comment]
Thanks! Ping @3rd-Eden if you need a new release.

**Comment by dickeylth** [comment]
@3rd-Eden I need a new release for this pr, can you help me? Thank you!

**Comment by lpinca** [comment]
cc: @Swaagie who can also publish on npm.

**Pull request merged**